File: find_common_characters/Find_Common_Characters.cpp
class Find_Common_Characters {
public:
    vector<string> commonChars(vector<string>& words) {
        if (words.empty()) {
            return vector<string>();
        }

        unordered_map<char, int> commons_counter;
        for (int i = 0; i < words[0].size(); i++) {
            commons_counter[words[0][i]]++;
        }

        for (int i = 1; i < words.size(); i++) {
            unordered_map<char, int> curr_counter;
            for (int j = 0; j < words[i].size(); j++) {
                curr_counter[words[i][j]]++;
            }

            for (auto& counter : commons_counter) {
                if (curr_counter.find(counter.first) == curr_counter.end()) {
                    counter.second = 0;
                    continue;
                }
                counter.second = min(counter.second, curr_counter[counter.first]);
            }
        }

        vector<string> common_chars;
        for (auto& counter : commons_counter) {
            for (int i = 0; i < counter.second; i++) {
                common_chars.push_back(string(1, counter.first));
            }
        }
        return common_chars;
    }
};

File: find_common_characters/FindCommonCharacters.java
class FindCommonCharacters {
    public List<String> commonChars(String[] words) {
        if (words == null || words.length < 2) {
            return new ArrayList<String>();
        }

        Map<Character, int[]> commonsCounter = new HashMap<>();
        for (int i = 0; i < words[0].length(); i++) {
            char ch = words[0].charAt(i);
            if (!commonsCounter.containsKey(ch)) {
                commonsCounter.put(ch, new int[]{0});
            }
            commonsCounter.get(ch)[0]++;
        }

        for (int i = 1; i < words.length; i++) {
            Map<Character, int[]> currCounter = new HashMap<>();
            for (int j = 0; j < words[i].length(); j++) {
                char ch = words[i].charAt(j);
                if (!currCounter.containsKey(ch)) {
                    currCounter.put(ch, new int[]{0});
                }
                currCounter.get(ch)[0]++;
            }

            for (Map.Entry<Character, int[]> counter : commonsCounter.entrySet()) {
                char ch = counter.getKey();
                if (!currCounter.containsKey(ch)) {
                    commonsCounter.get(ch)[0] = 0;
                    continue;
                }
                commonsCounter.get(ch)[0] = Math.min(counter.getValue()[0], currCounter.get(ch)[0]);
            }
        }

        List<String> commonChars = new ArrayList<>();
        for (Map.Entry<Character, int[]> counter : commonsCounter.entrySet()) {
            for (int i = 0; i < counter.getValue()[0]; i++) {
                commonChars.add(String.valueOf(counter.getKey()));
            }
        }
        return commonChars;
    }
}